Deactivation notification

rajivnathan opened the pull request:
This PR adds code to send an email to the user's account when their account has been deactivated. This applies to both when their account is manually or automatically deactivated.

Related issue: https://issues.redhat.com/browse/CRT-644

e2e tests: https://github.com/codeready-toolchain/toolchain-e2e/pull/165

alexeykazakov commented on pkg/controller/usersignup/usersignup_controller.go:
```diff
@@ -813,6 +819,33 @@ func (r *ReconcileUserSignup) updateStatusConditions(userSignup *toolchainv1alph
+	}
+
+	if err := r.client.Create(context.TODO(), notification); err != nil {
+		logger.Error(err, "Failed create deactivation notification resource")
```
Any tests for this error?

rajivnathan replied:
Included in https://github.com/codeready-toolchain/host-operator/pull/289/commits/ba8ad1c44e6ef9e9e86ac3c84d5da84be6ef3300

rajivnathan commented on pkg/controller/usersignup/usersignup_controller.go:
```diff
@@ -324,7 +326,11 @@ func (r *ReconcileUserSignup) Reconcile(request reconcile.Request) (reconcile.Re
-		return reconcile.Result{}, r.updateStatus(reqLogger, instance, r.setStatusDeactivated)
+		err := r.updateStatus(reqLogger, instance, r.setStatusDeactivated)
+		if err == nil {
+			err = r.sendDeactivatedNotification(reqLogger, instance)
```
Oh right, forgot about that. Will add it.

rajivnathan replied:
Included in https://github.com/codeready-toolchain/host-operator/pull/289/commits/ba8ad1c44e6ef9e9e86ac3c84d5da84be6ef3300

rajivnathan commented:
@alexeykazakov I updated the controller to set a new UserSignup status called `UserDeactivatedNotificationCreated` that has 3 states: `NotificationCRCreated`, `NotificationCRNotCreated` and `NotificationCRCreationFailed`. 
`NotificationCRCreated` is set when UserSignup.Spec.Deactivated = `true` and the notification resource was successfully created
`NotificationCRCreationFailed` is set when UserSignup.Spec.Deactivated = `true` and we failed to create the notification resource
`NotificationCRNotCreated` is set when UserSignup.Spec.Deactivated = `false`

I also added some tests for the reactivating a UserSignup after deactivation scenario.

alexeykazakov commented on pkg/controller/usersignup/usersignup_controller.go:
```diff
@@ -128,6 +125,13 @@ func (r *ReconcileUserSignup) Reconcile(request reconcile.Request) (reconcile.Re
 
+	// If the usersignup is deactivated then ensure the deactivated notification status is set to false, this is especially important for cases when a user is deactivated and then reactivated
+	if !instance.Spec.Deactivated {
+		if err := r.updateStatus(reqLogger, instance, r.setStatusDeactivationNotificationNotCreated); err != nil {
```
Hm.. yeah. I didn't think about that but we do need to cover that case indeed. Good catch.
The only thing is that I would probably use a different reason. The CR is not created because it's not created doesn't sound super informative :)
I would use "UserIsActive" or "UserIsNotDeactivated" as the reason for the "NotificationCreated/False" condition :)

rajivnathan replied:
Haha, yeah I guess it's not super informative. I changed it to `UserIsActive`

included in https://github.com/codeready-toolchain/host-operator/pull/289/commits/6bf0e589a480fce21f3bd1e4c69414ff0c73b23f

alexeykazakov commented on pkg/controller/usersignup/usersignup_controller_test.go:
```diff
@@ -1332,6 +1423,212 @@ func TestUserSignupDeactivatedAfterMURCreated(t *testing.T) {
+			})
+		AssertThatCounterHas(t, 2)
+
+		// A deactivated notification should have been created
```
It should or it shouldn't? You assert that it's not created.

rajivnathan replied:
Oops, should be "shouldn't". Fixed in https://github.com/codeready-toolchain/host-operator/pull/289/commits/448e505672962faf3eda38deaa43deb1de12d8c7

alexeykazakov commented on pkg/controller/usersignup/usersignup_controller_test.go:
```diff
@@ -1332,6 +1423,212 @@ func TestUserSignupDeactivatedAfterMURCreated(t *testing.T) {
+			})
+		AssertThatCounterHas(t, 2)
+
+		// A deactivation notification should not be created because this is the reactivation case
```
I'm a bit lost what you are actually testing here?
Is it a test for covering status update case? Then what does the reactivation have with it? I don't even see that the user is reactivated in the test.

rajivnathan replied:
When a usersignup is deactivated (`Deactivated: true`) then it should get status condition `NotificationCRCreated`. So the test starts with a usersignup has `Deactivated: false` but it has condition `NotificationCRCreated` so this is the scenario where a user was deactivated and was just reactivated. Then the reconcile should update the status to `UserIsActive`. I added comments and initialized the status conditions in the individual test functions to make it a bit more clear. Does it help? Any other suggestions?

alexeykazakov replied:
OK. Makes sense. The comments helped but I think we can make them even clearer :) See my suggestions.

alexeykazakov commented on pkg/controller/usersignup/usersignup_controller_test.go:
```diff
@@ -1332,6 +1423,212 @@ func TestUserSignupDeactivatedAfterMURCreated(t *testing.T) {
+		AssertThatCounterHas(t, 3)
+
+		// There should not be a notification created because the user was reactivated
+		notifications := &v1alpha1.NotificationList{}
```
Same here. It doesn't seem to be testing what you state. We do need to test the reactivation flow but it seems to be missing.

rajivnathan replied:
Same as above comment

alexeykazakov commented on pkg/controller/usersignup/usersignup_controller_test.go:
```diff
@@ -1315,8 +1407,251 @@ func TestUserSignupDeactivatedAfterMURCreated(t *testing.T) {
+
+	t.Run("when reactivating the usersignup successfully", func(t *testing.T) {
+		// given
+		// start with a usersignup that has the Notification Created status which signals an active user
```
```suggestion
		// start with a usersignup that has the Notification Created status set to "true" but Spec.Deactivated is set to "false" which signals a user which has been just reactivated.
```
Right?

alexeykazakov commented on pkg/controller/usersignup/usersignup_controller_test.go:
```diff
@@ -1315,8 +1407,251 @@ func TestUserSignupDeactivatedAfterMURCreated(t *testing.T) {
+
+	t.Run("when resetting the usersignup deactivation notification status fails", func(t *testing.T) {
+		// given
+		// start with a usersignup that has the Notification Created status which signals an active user
```
```suggestion
		// start with a usersignup that has the Notification Created status set to "true" but Spec.Deactivated is set to "false" which signals a user which has been just reactivated.
```

alexeykazakov reviewed (approved):
Looks good! Just a couple of suggestions to improve code readability.

MatousJobanek commented on pkg/controller/usersignup/usersignup_controller_test.go:
```diff
@@ -1384,6 +1704,39 @@ func TestUserSignupDeactivatingWhenMURExists(t *testing.T) {
+		require.NoError(t, err)
+		require.Len(t, notifications.Items, 0)
+
+		// 2nd reconcile should handle creating the deactivation notification
+		res, err := r.Reconcile(req)
+		require.NoError(t, err)
+		require.Equal(t, reconcile.Result{}, res)
+
+		// lookup the userSignup and check the conditions
+		err = r.client.Get(context.TODO(), types.NamespacedName{Name: userSignup.Name, Namespace: req.Namespace}, userSignup)
+		require.NoError(t, err)
+
+		// Confirm the status has been set to Deactivated and the deactivation notification is created
+		test.AssertConditionsMatch(t, userSignup.Status.Conditions,
+			v1alpha1.Condition{
+				Type:   v1alpha1.UserSignupApproved,
+				Status: v1.ConditionTrue,
+				Reason: "ApprovedAutomatically",
+			},
+			v1alpha1.Condition{
+				Type:   v1alpha1.UserSignupComplete,
+				Status: v1.ConditionTrue,
+				Reason: "Deactivated",
+			},
+			v1alpha1.Condition{
+				Type:   v1alpha1.UserSignupUserDeactivatedNotificationCreated,
+				Status: v1.ConditionTrue,
+				Reason: "NotificationCRCreated",
+			})
```
Could you please move this into a subtest? This verifies another functionality based on the previous test - it's much easier to read when it's used as a subtest.

rajivnathan replied:
Sure, makes sense!

rajivnathan replied:
Done in https://github.com/codeready-toolchain/host-operator/pull/289/commits/a546d2527c71eb6962eb1428bc6ec22e8f6ceca9

MatousJobanek commented on pkg/controller/usersignup/usersignup_controller_test.go:
```diff
@@ -1679,6 +2037,14 @@ func TestUserSignupDeactivatedButMURDeleteFails(t *testing.T) {
 	AssertThatCounterHas(t, 1)
+
+	// There should not be a notification created since the mur deletion failed even if reconciled again
+	_, err = r.Reconcile(req)
+	require.Error(t, err)
+	notifications := &v1alpha1.NotificationList{}
+	err = r.client.List(context.TODO(), notifications)
+	require.NoError(t, err)
+	require.Len(t, notifications.Items, 0)
```
I believe that these kinds of checks could be extracted to a separated function - it makes the code more compact and shorter.

rajivnathan replied:
Good suggestion, created a new function in the test/notification package for that. Done in https://github.com/codeready-toolchain/host-operator/pull/289/commits/c53daf95d071e6c1317f7f9e1ad78d0fd163e927

MatousJobanek reviewed (approved):
Good job. I really like the solution where the deactivation notification is managed by UserSignup controller based on the fact that the MUR was deleted :+1:

rajivnathan commented:
/retest
